File: 2021/js/lib/day01.js
'use strict';

const run = (s) => {
    const input = s.split('\n').map(s => parseInt(s));
    console.log(`day01 part 1 => ${part1(input)}`); // 1374
    console.log(`day01 part 2 => ${part2(input)}`); // 1418
};

/**
 * @example part1([ 199, 200, 208, 210, 200, 207, 240, 269, 260, 263 ])
 * //=> 7
 */
const part1 = (input) => {
    let count = 0;
    for (let i = 1; i < input.length; i++) {
        if (input[i - 1] < input[i]) {
            count = count + 1;
        }
    }
    return count;
};

/**
 * @example part2([ 199, 200, 208, 210, 200, 207, 240, 269, 260, 263 ])
 * //=> 5
 */
const part2 = (input) => {
    let count = 0;
    for (let i = 2; i < input.length; i++) {
        if ((input[i - 3] + input[i - 2] + input[i - 1]) <
            (input[i - 2] + input[i - 1] + input[i])) {
            count = count + 1;
        }
    }
    return count;
};

module.exports = {
    run,
    part1,
    part2
};

File: 2021/erlang/src/day01.erl
-module(day01).

-export([run/1]).

run(Input) ->
    Numbers = lists:map(fun list_to_integer/1, Input),
    io:format("day01 part 1 => ~p~n", [part1(Numbers)]),
    io:format("day01 part 2 => ~p~n", [part2(Numbers)]).

part1(Numbers) -> part1(Numbers, 0).

part1(L, Acc) when length(L) =:= 1 -> Acc;
part1([F, S | Rest], Acc) when F < S -> part1([S|Rest], Acc + 1);
part1([F, S | Rest], Acc) when F > S -> part1([S|Rest], Acc).

part2(Numbers) -> part2(Numbers, 0).

part2(L, Acc) when length(L) =:= 3 ->
    Acc;
part2([Fst, Snd, Thd, Fth | Rest], Acc) ->
    case ((Fst + Snd + Thd) < (Snd + Thd + Fth)) of
        true -> part2([Snd|[Thd|[Fth|Rest]]], Acc +1);
        false -> part2([Snd|[Thd|[Fth|Rest]]], Acc)
    end.
